### ISSUE TITLE
External HostFilter implementations cannot be tested against HostFilter interface

### ISSUE THREAD

**Issue by jacobgreenleaf** [gocql/gocql#1311]
Please answer these questions before submitting your issue. Thanks!

### What version of Cassandra are you using?
n/a

### What version of Gocql are you using?
master

### What did you do?
```
// hostfilter.go
type datacenterRegexFilter struct { pattern *regexp.Regexp }
func (d *datacenterRegexFilter) Accept(host *gocql.HostInfo) bool {
    return d.pattern.MatchString(host.DataCenter())
}
var _ gocql.HostFilter = (*datacenterRegexFilter)(nil)

// hostfilter_test.go
func TestDatacenterRegexHostFilter(t *testing.T) {
    f := &datacenterRegexFilter{pattern: regexp.MustCompile("us-east-1a|us-east-1d")}
    assert.True(t, f.Accept(&gocql.HostInfo{
        // dataCenter: "us-east-1a"
    }))
}
```

### What did you expect to see?
Test is successful

### What did you see instead?
Because `dataCenter` is unexported and there is no setter, this test cannot be written. The [tests for other host filters](https://github.com/gocql/gocql/blob/master/filters_test.go#L20) can be written because they create new `*HostInfo` structs inline by assigning to unexported fields. But implementations that use the `HostFilter` interface to provide a custom filter cannot be tested without boilerplate code. 

Host filters that only filter on the connection address can use the public setter [`SetConnectAddress`](https://godoc.org/github.com/gocql/gocql#HostInfo.SetConnectAddress). However filters that use other fields like `dataCenter` cannot use the unexported setters like `setDataCenter`.

**Comment by alourie** [comment]
@jacobgreenleaf we have the `DataCentreHostFilter` call which returns a HostFilter which filters by a specific DataCentre. Would it help if we expanded it (or added a new function) to support regexp filtering? I don't mind exporting a bunch of setters, but if they're not used directly and only needed for _client side_ testing, I'm quite hesitant to do so.

**Comment by jacobgreenleaf** [comment]
> @jacobgreenleaf we have the `DataCentreHostFilter` call which returns a HostFilter which filters by a specific DataCentre. Would it help if we expanded it (or added a new function) to support regexp filtering? I don't mind exporting a bunch of setters, but if they're not used directly and only needed for _client side_ testing, I'm quite hesitant to do so.

Adding that would allow this use case but would not make the public interface of `HostFilter` more friendly. I imagine that the purpose of the `HostFilter` interface is to allow users to customize the host selection policy with implementations that you don't or won't provide or that use code that can't be moved into gocql for whatever reason. In that case it would be nice to have a way to test the host selection policy implementations. You might compare it to something like `http.Handler` where the `*http.Request` struct has public fields and also provides a public constructor `http.NewRequest`. 

If `HostFilter` is intended to be only be used with `gocql` native implementations (`AcceptAllFilter`, `DenyAllFilter`, `DataCentreHostFilter` and `WhiteListHostFilter`) then this is not an issue.

**Comment by OleksiienkoMykyta** [comment]
I think it would be a useful enhancement to allow a user to customize host selection policy from the box without any workarounds. I would say it's pretty relevant.
@joao-r-reis, what do you make of it?

**Comment by joao-r-reis** [comment]
I agree, I think the solution would be to create a new interface that `HostInfo` implements and use it in `Accept()` so that apps can mock the interface for their own tests

**Comment by joao-r-reis** [comment]
Created [CASSGO-71](https://issues.apache.org/jira/browse/CASSGO-71)

**Comment by joao-r-reis** [comment]
Fixed in #1892
